### PR TITLE
refactor!: Convert AddContentApi to Kotlin

### PR DESCRIPTION
**Pull request by david-allison**
Closes https://github.com/ankidroid/Anki-Android/pull/11377

Converted with disabled git hooks + tested manually

**The conversion caused breaking changes WRT nullability of parameters AND changing `fun` to `val`**

**This is NOT the final nullability which we desire, we perform the conversion in a single commit to reduce the chance of bugs being introduced**

----

Changes from a 'pure' conversion':

- addNotes:
  - fieldsList: changed to `List<Array<String>?>`
  - tagsList: changed to `List<Set<String>?>?`

- addNewCustomModel changed from auto conversion:
  - fields: -> `Array<String>`
  - cards -> `Array<String>`
  - qfmt -> `Array<String>`
  - afmt -> `Array<String>`

- `splitFields(flds).length` -> `splitFields(flds).size`

fix name shadowing:
  - `var preferredName = preferredName`

add `!!`
  - `val fname = File(returnUri!!.path).toString()`
  - `val fname = File(returnUri!!.path!!).toString()`

remove `private`:
  - `private get() = if (apiHostSpecVersion < 2) CompatV1() else CompatV2()`
  - `get() = if (apiHostSpecVersion < 2) CompatV1() else CompatV2()`

remove imports:
 - NoteInfo.Companion.buildFromCursor
 - Utils.fieldChecksum
 - Utils.joinFields
 - Utils.joinTags
 - Utils.splitFields

add imports: FlashCardsContract
 - Card
 - CardTemplate
 - Deck
 - Model
 - Note

many spacing changes:
  - sample: `values.put(Note.TAGS, Utils.joinTags(tagsList[i]))`
    - moved back to a single line

Moved api docs from `<pre>` to \`\`\`

### CODE REVIEW

**Comment by david-allison** [comment]
High priority: Kotlin Cleanups need to be performed and the `api` module needs a thorough audit before release

File: api/src/main/java/com/ichi2/anki/api/AddContentApi.kt [hunk 1]
```diff
@@ -140,108 +111,104 @@ public final class AddContentApi {
...
-     * <pre>
-     *     <code>
-     *     getContext().grantUriPermission("com.ichi2.anki", uri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
-     *     // Then when file is added, remove the permission
-     *     // add File ...
-     *     getContext().revokePermission(uri, Intent.FLAG_GRAN_READ_URI_PERMISSION)
-     *     </code>
-     * </pre>
+     * ```
+     * getContext().grantUriPermission("com.ichi2.anki", uri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
```
Comment (david-allison) [inline api/src/main/java/com/ichi2/anki/api/AddContentApi.kt:142]: Do we want the indent back?

Comment (lukstbit) [reply api/src/main/java/com/ichi2/anki/api/AddContentApi.kt:142]: Absolutely yes, right now it's just one big piece of text with low readability.

File: api/src/main/java/com/ichi2/anki/api/AddContentApi.kt [hunk 2]
```diff
@@ -632,176 +581,248 @@ public final class AddContentApi {
...
+        )
+
+        /**
+         * Get the AnkiDroid package name that the API will communicate with.
+         * This can be used to check that a supported version of AnkiDroid is installed,
+         * or to get the application label and icon, etc.
+         * @param context a Context that can be used to get the PackageManager
+         * @return packageId of AnkiDroid if a supported version is not installed, otherwise null
+         */
+        fun getAnkiDroidPackageName(context: Context): String? {
```
Comment (lukstbit) [inline api/src/main/java/com/ichi2/anki/api/AddContentApi.kt:816]: // TODO simplify into one line fun

**Review by lukstbit — CHANGES_REQUESTED** [review]
Looks good with minor todos and the doc issue.

File: api/src/main/java/com/ichi2/anki/api/AddContentApi.kt [hunk 3]
```diff
@@ -140,108 +111,104 @@ public final class AddContentApi {
...
-
-        try {
-            Uri returnUri = mResolver.insert(AnkiMedia.CONTENT_URI, contentValues);
+    fun addMediaFromUri(
+        fileUri: Uri,
+        preferredName: String,
+        mimeType: String
+    ): String? {
+        @Suppress("NAME_SHADOWING")
+        var preferredName = preferredName
```
[the remark this answers is not in the document]
Comment (david-allison) [reply api/src/main/java/com/ichi2/anki/api/AddContentApi.kt:172]: Only adding `@Suppress("NAME_SHADOWING")`, will update the comment

**Review by lukstbit — APPROVED** [review]
Lets get this in!

**Review by BrayanDSO — APPROVED** [review]
boom (haven't left nullability cleanups because there is an issue for it)

**Pull request merged**